SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetConnection.cpp

```diff
@@ -193,3 +196,114 @@ void USpatialNetConnection::OnHeartbeat()
+
+				Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(Op.update.schema_type);
+				uint32 EventCount = Schema_GetObjectCount(EventsObject, SpatialConstants::PING_EVENT_ID);
+
+				if (EventCount > 0 && ConnectionPtr->PlayerController)
+				{
+					if (APlayerState* PlayerState = Cast<APlayerController>(ConnectionPtr->PlayerController)->PlayerState)
+					{
+						uint32 ID = Schema_GetUint32(EventsObject, SpatialConstants::PING_ID_OFFSET_ID);
+						float fTimeSent;
+						if (ConnectionPtr->SentPingTimestamps.RemoveAndCopyValue(ID, fTimeSent))
+						{
+							// Set player state ExactPing in msecs, ExactPing is not replicated
+							PlayerState->ExactPing = (fReceivedTimestamp - fTimeSent) * 1000.0f;
+
+							// In native Unreal, Ping property on player state is replicated
+							// (compressed for replication by dividing ExactPing by 4)
+							PlayerState->Ping = (int32)(PlayerState->ExactPing * 0.25f);
+						}
+					}
+				}
+			}
+		}));
+
+		TimerManager->SetTimer(PingTimer, [this]()
+		{
+			RemoveExpiredPings();
+			SendPingOrPong(CurrentPingID++, SpatialConstants::SERVER_PING_COMPONENT_ID);
+		}, GetDefault<USpatialGDKSettings>()->PingIntervalSeconds, true, 0.0f);
+	}
+	else
+	{
+		Cast<USpatialNetDriver>(Driver)->Receiver->AddServerPingDelegate(PlayerControllerEntity, ServerPingDelegate::CreateLambda([ConnectionPtr](Worker_ComponentUpdateOp& Op)
+		{
+			if (ConnectionPtr.IsValid())
+			{
+				Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(Op.update.schema_type);
+				uint32 EventCount = Schema_GetObjectCount(EventsObject, SpatialConstants::PING_EVENT_ID);
+
+				if (EventCount > 0)
+				{
+					uint32 ID = Schema_GetUint32(EventsObject, SpatialConstants::PING_ID_OFFSET_ID);
+					ConnectionPtr->SendPingOrPong(ID, SpatialConstants::CLIENT_PONG_COMPONENT_ID);
+				}
+			}
+		}));
+	}
+}
+
+void USpatialNetConnection::DisablePing()
+{
+	if (TimerManager && PingTimer.IsValid())
+	{
+		TimerManager->ClearTimer(PingTimer);
+	}
+	SentPingTimestamps.Empty();
+
+	PlayerControllerEntity = SpatialConstants::INVALID_ENTITY_ID;
+}
+
+void USpatialNetConnection::SendPingOrPong(uint32 PingId, Worker_ComponentId ComponentId)
+{
+	Worker_ComponentUpdate ComponentUpdate = {};
+
+	ComponentUpdate.component_id = ComponentId;
+	ComponentUpdate.schema_type = Schema_CreateComponentUpdate(ComponentId);
+	Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(ComponentUpdate.schema_type);
+	Schema_AddObject(EventsObject, SpatialConstants::PING_EVENT_ID);
+	Schema_AddUint32(EventsObject, SpatialConstants::PING_ID_OFFSET_ID, PingId);
+
+	USpatialWorkerConnection* Connection = Cast<USpatialNetDriver>(Driver)->Connection;
+	if (Connection->IsConnected())
+	{
+		Connection->SendComponentUpdate(PlayerControllerEntity, &ComponentUpdate);
+
+		if (Driver->IsServer())
+		{
+			SentPingTimestamps.Add(PingId, GetWorld()->GetTimeSeconds());
+		}
+	}
+}
+
+void USpatialNetConnection::RemoveExpiredPings()
```

Comment:
Why do we need this? Can't this cause a segfault when we eventually receive it? Isn't the point in using events that all pings get received? The client won't know which pings to ignore so it will respond to all of them.

Reply:
If a client doesn't respond we should also set the ping to some maximum.

Reply:
Removed